### PR TITLE
Allow caution payments for invoices

### LINKED ISSUE #808: (feat) Handling automatically the CAUTION when invoicing a patient
This issue is about to handle CAUTION of a patient automatically when invoicing the patient.

We must have a mean to signal to the user of BHIMA who makes invoice that the patient has a CAUTION and if we have to use this CAUTION for an invoice.

### Caution Steps: 

1. Client pay a caution via the `cash payment module`

    - The client will be `credited`
    - The cashbox will be `debited`
    - A caution receipt is generated as proof 

2. Bill the patient : 

    - Display the remaining caution amount, in reality the balance of the client which is `< 0`
    - In this step, the caution displayed is for `just notify` the user that the enterprise must money to the client
    - The caution amount doesn't affect the invoice, invoices are independent to cautions

3. Patient Financial Activities : 

    - Notify also for a remaining caution amount of the client

### PR DESCRIPTION
**Pull request by jniles**
This PR presents the initial step in supporting cautions/prepayments in BHIMA.  The following components have been implemented:

 1. A reliable calculation of the patient's balance with the enterprise.  If a patient has a creditor balance with the hospital, the Patient Invoice module will signal the user that the patient has a creditor balance with an alert.
 2. Automatic "linking" of the patient's previous cash prepayment/caution when they are invoiced.  This is done by creating a third transaction that both debits and credits the same accountant.  Interestingly, this rule is not caught by the Trial Balance, which it probably should.

This PR is now complete!  It includes the option of toggling on/off prepayments, which will be particularly useful to institutions that do not accept prepayments.  The following issues have been closed:

 1. Closes #808 
 2. Closes https://github.com/IMA-WorldHealth/bhima-2.X/issues/1324
 3. Closes https://github.com/IMA-WorldHealth/bhima-2.X/issues/2687
 4. Closes https://github.com/IMA-WorldHealth/bhima-2.X/issues/2695

### CODE REVIEW

**Comment by jniles** [comment]
@mbayopanda, can you give this a quick review?  I'd like to break this into smaller components, so that you don't have a ton to review at the end.

File: server/models/procedures/invoicing.sql [hunk 1]
```diff
@@ -17,8 +16,7 @@ CREATE PROCEDURE StageInvoice(
   IN date DATETIME,
   IN cost DECIMAL(19, 4) UNSIGNED,
   IN description TEXT,
-  IN service_id SMALLINT(5) UNSIGNED,
-  IN debtor_uuid BINARY(16),
+  IN service_id SMALLINT(5) UNSIGNED, IN debtor_uuid BINARY(16),
```
Comment (mbayopanda) [inline server/models/procedures/invoicing.sql]: Why this in one line instead of multiple lines.

File: server/models/procedures/invoicing.sql [hunk 2]
```diff
@@ -580,5 +578,196 @@ BEGIN
...
+  DECLARE enterpriseCurrencyId SMALLINT(5);
+  DECLARE linkTransactionTypeId SMALLINT(5);
+  DECLARE debtorAccountId INT(10);
+
+
+  DECLARE curse CURSOR FOR
+    SELECT payment.uuid, payment.balance FROM stage_payment_balances AS payment;
+  DECLARE CONTINUE HANDLER FOR NOT FOUND SET done = TRUE;
+
+  CALL SetupDebugLog();
```
Comment (mbayopanda) [inline server/models/procedures/invoicing.sql]: Is this call necessary ?

File: sh/build-database.sh [hunk 3]
```diff
@@ -29,8 +29,9 @@ echo "[build] functions"
 mysql -u $DB_USER -p$DB_PASS $DB_NAME < server/models/functions.sql &> /dev/null
 
 echo "[build] procedures"
-mysql -u $DB_USER -p$DB_PASS $DB_NAME < server/models/procedures.sql &> /dev/null
+mysql -u $DB_USER -p$DB_PASS $DB_NAME < server/models/procedures.sql
 mysql -u $DB_USER -p$DB_PASS $DB_NAME < server/models/admin.sql &> /dev/null
+mysql -u $DB_USER -p$DB_PASS $DB_NAME < server/models/debug.sql &> /dev/null
```
Comment (mbayopanda) [inline sh/build-database.sh]: `debug.sql` is now a part of development environment.

File: server/models/procedures/invoicing.sql [hunk 4]
```diff
@@ -580,5 +578,196 @@ BEGIN
...
+
+  -- TODO(@jniles) - decide on this for real
+  SET linkTransactionTypeId = 3;
```
Comment (mbayopanda) [inline server/models/procedures/invoicing.sql]: `TransactionTypeId = 3` is for convention payment, so maybe we can add a new transaction type for Caution

caution link transaction type

**Review by mbayopanda — COMMENTED** [review]
I made a comment on `transactionType=3` apart from that the caution payment and its link to invoice work well.

**Comment by mbayopanda** [comment]
@jniles ping me when it will be possible to `bors` this PR

File: server/models/procedures/invoicing.sql (hunk 2 above)
Comment (jniles) [reply server/models/procedures/invoicing.sql]: Hehehe nope.  Good catch.

**Comment by jniles** [comment]
@mbayopanda, this is ready for another review.

File: client/src/modules/cash/cash-form.service.js [hunk 5]
```diff
@@ -29,10 +29,16 @@ function CashFormService(AppCache, Session, Patients, Exchange) {
   CashForm.prototype.setup = function setup() {
     this.details = { date : new Date(), description : '', invoices : [] };
     this.details.currency_id = Session.enterprise.currency_id;
+    this.hasPrepaymentSupport = Session.enterprise.settings.enable_prepayments;
```
Comment (mbayopanda) [inline client/src/modules/cash/cash-form.service.js:32]: 👍

**Review by mbayopanda — APPROVED** [review]
LGTM

Here are some things that i noticed : 
- In the enterprise module we can have a button for updating enterprise information in the section "Parametre de l'entreprise" for an easy use

- Translate are missing in the message of caution in the invoicing page.

**Comment by jniles** [comment]
@mbayopanda, thanks for the review!

I'll put in a patch with the translations soon.

**Pull request merged**